rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,223 @@
+
+We could use the occasion and introduce more operators like those mentioned above.
+
+Function composition is mostly interesting for the so-called "point-free" programming style,
+where partially applied compositions of functions are preferred over the introduction of lambda terms.
+However, Nix is not well suited for that programming style for various reasons,
+nor would that point-free style have nearly as many applications in real-world Nixpkgs code.
+
+F#'s reverse-pipe operator has a lot less use due to its left-associativity,
+but a right-associative version of it more similar to Haskell's `$` might be an alternative:
+
+```nix
+defaultPrefsFile = 
+  pkgs.writeText "nixos-default-prefs.js" <|
+  lib.concatStringsSep "\n" <|
+  lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  )
+  defaultPrefs
+  ;
+```
+
+Adding both pipe directions raises questions about how these two interact when used together.
+F# has the same binding strength for both, but this only works well because both are left-associative.
+Haskell has `&` stronger than `$`, which is very sensible but unlikely to be intuitive to a new user.
+Given that we want to call them `|>` and `<|` instead, then users might equally well to assume both have
+equal strength.
+
+Given these restrictions and the fact that situations where one needs both in Nix are expected to be fairly rare,
+it is recommended to choose either one of `|>` and `<|`, but not have both in the language.
+
+## Change the `pipe` function signature
+
+There are many equivalent ways to declare this function, instead of just using the current design.
+For example, one could flip its arguments to allow a partially-applied point-free style (see above).
+One could also make this a single-argument function so that it only takes the list as argument.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Introducing `|>` has the drawback of adding complexity to the language, and it will break older tooling.
+- The main purpose of `builtins.pipe` is as a stop-gap until Nixpkgs can use `|>`. After that, it will be mostly redundant.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Who is going to implement this in Nix?
+- How difficult will the implementation be?
+- Will this affect evaluation performance in some way?
+  - There is reason to expect that replacing `lib.pipe` with a builtin will reduce its overhead,
+    and that the builtin should have little to no overhead compared to regular function application.
+
+# Future work
+[future]: #future-work
+
+Once introduced and usable in Nixpkgs, existing code may benefit from being migrated to using these features.
+Automatically transforming nested function calls into pipelines is unlikely,
+as doing so is not guaranteed to always be a subjective improvement to the code.
+It might be possible to write a lint which detects opportunities for piping, for example in nixpkgs-hammering.
```

Comment:
Overly zealous rules like this make people focus on the arbitrary rather than solving the actual problems.
You could fill your day bothering people with style suggestions, but that's not going to help anyone.
We have real unsolved problems in Nixpkgs and NixOS. How about getting out of your comfort zone and actually achieve something that solves a real packaging/deployment problem? (Not talking about this RFC necessarily, but about style rule policing)

Reply:
I'm not saying that we should do this, I'm just stating the possibility that we *could*.